src/core/net/ip6.cpp

```diff
@@ -1121,7 +1121,8 @@ Error Ip6::HandleDatagram(OwnedPtr<Message> aMessagePtr, const void *aLinkMessag
         }
 #endif
 
-        forwardHost = header.GetDestination().IsMulticastLargerThanRealmLocal();
+        // Forward multicast packets to host network stack unless they came from there.
+        forwardHost = !aMessagePtr->IsOriginHostUntrusted();
```

Comment:
The existing code would not apply any check on origin. So we would have forwarded a message originating from host back to host when `GetDestination().IsMulticastLargerThanRealmLocal()` and now we wont? Should we keep this behavior unchanged?

I think the main change we want is to allow all multicast traffic to be passed to host (not just `IsMulticastLargerThanRealmLocal()`), so how about this?
```suggestion
        forwardHost = true;
```

Reply:
@librasungirl  I'm afraid passing a packet from host back to host would cause infinite loop or duplicate packets being forwarded. Do we have a explicit use case for delivering packets back to the host?

Reply:
We have APIs that allow caller to specify the "multicast loop" desired behavior (per messages) -> `otMessageSetLoopbackToHostAllowed()`. 

> Do we have a known use case for delivering packets back to the host?


I don't know about exact use-cases but I expect there may be some. 
- There were some discussions on this specific behavior here -> https://github.com/openthread/openthread/issues/9489
- So I suggest not changing existing behavior, in particular regarding how the "larger than realm local" multicast msg from host are handled. Today, we allow them to be passed back to host (assuming `otMessageSetLoopbackToHostAllowed` is set).

Reply:
For `multicast loop`, I think the host stack should handle it by enabling the `IPV6_MULTICAST_LOOP` on the socket that desires the behavior. We don't need to pass back a packet to the host again, as it may conflict with the host stack configuration. Thoughts?

Reply:
I checked `otMessageSetLoopbackToHostAllowed()` is currently used in `src/posix/platform/netif.cpp`. @superwhd @librasungirl @wgtdkp is this necessary? Actually I suspect this would result in duplicate messages (scope larger than realm local) being processed by sockets in the host.

Reply:
> For multicast loop, I think the host stack should handle it by enabling the IPV6_MULTICAST_LOOP on the socket that desires the behavior. 

I guess rhis can work for posix, but  OT may be  used with other OS/stacks. The https://github.com/openthread/openthread/issues/9489 talks about Zephyr and multicast used. I suggested a similar suggestion to what you have in mind here in this comment https://github.com/openthread/openthread/issues/9489#issuecomment-1748434030 but the API is added now to support this.